Don't emit the FORCE_LOAD symbols when the compiler is running

jimingham opened the pull request:
on behalf of the debugger.  The debugger will read the LinkLibrary's
from all the modules it sees, and hand load all the required dependencies,
and since the symbol is weak it doesn't even tell us whether a
required dependency is missing. So it serves no purpose in this case.

<rdar://problem/51463642>

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

jimingham committed:
Don't emit the FORCE_LOAD symbols when the compiler is running
on behalf of the debugger.  The debugger will read the LinkLibrary's
from all the modules it sees, and hand load all the required dependencies,
and since the symbol is weak it doesn't even tell us whether a
required dependency is missing. So it serves no purpose in this case.

<rdar://problem/51463642>

jimingham commented:
@swift-ci please test

jrose-apple commented on lib/IRGen/IRGenModule.cpp:
```diff
@@ -1033,7 +1033,12 @@ void IRGenModule::addLinkLibrary(const LinkLibrary &linkLib) {
+  // from all the modules it sees, and hand load all the required dependencies,
+  // and since the symbol is weak it doesn't even tell us whether a
+  // required dependency is missing. So it serves no purpose in this case.
+  if (linkLib.shouldForceLoad() && !Context.LangOpts.DebuggerSupport) {
```
A belated thought: given that the debugger is going to read the `LinkLibrary`s itself, maybe it shouldn't even bother generating the metadata above either. You could just bail out at the top of the function.

jimingham replied:
I'll give that a try and see if lldb and the REPL are still happy.

jimingham replied:
That also works, and lldb seems happy.

jimingham committed:
We don't need the metadata for the auto linking when debugging either.

So it is simpler to just early return when we are running on the
debugger's behalf.

jimingham commented:
@swift-ci please test

jrose-apple commented on lib/IRGen/IRGenModule.cpp:
```diff
@@ -1009,6 +1009,15 @@ llvm::SmallString<32> getTargetDependentLibraryOption(const llvm::Triple &T,
+  // from all the modules it sees, and hand load all the required dependencies.
+  // so it doesn't need this information.  And since the FORCE_LOAD symbol is
+  // weak it doesn't even tell us whether a required dependency is missing.
+  // So it serves no purpose in this case.
```
Looks good, but the comment is no longer accurate. I think it can just be "The debugger can get autolink information directly from the LinkLibraries of the module, so there's no reason to generate it in the LLVM IR."

jimingham committed:
Fix the comment to be more concise.

jimingham commented:
@swift-ci please test

jimingham commented:
@jrose-apple: Does that look better?

jrose-apple reviewed (approved):
Yep. Thanks for tracking this down!

jimingham commented:
@swift-ci please test

jimingham committed:
Fix the valance of the DebuggerSupport test.

I was moving this around per Jordan's request and didn't change
the valance of the if test.

jimingham commented:
@swift-ci please test

jimingham commented:
@swift-ci please test
